smolyay/basis.py

```diff
@@ -335,17 +335,23 @@ def derivative(self, x):
         ValueError
             if input is outside the domain [-1,1].
         """
-        if x > 1 or x < -1:
+        if numpy.any([numpy.greater(x,1),numpy.less(x,-1)]):
             raise ValueError("Input is outside the domain [-1,1]")
-        if x == 1:
-            return self._n*(self._n+1)*(self._n+2)/3
-        elif x == -1:
-            return ((-1)**(self._n+1))*self._n*(self._n+1)*(self._n+2)/3
+        if self._derivative_polynomial is None:
+            self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
+        x = numpy.array(x, copy=False)
+        y = numpy.zeros(x.shape)
+        flag1 = numpy.logical_not(numpy.logical_or(x == 1,x == -1))
+        flag2 = x == 1
+        flag3 = x == -1
+        y[flag1] =  ((self._n+1)*self._derivative_polynomial(x[flag1]) -
+                     x[flag1]*self(x[flag1]))/(x[flag1]**2-1)
+        y[flag2] = self._n*(self._n + 1)*(self._n + 2)/3
+        y[flag3] = ((-1)**(self._n+1))*self._n*(self._n + 1)*(self._n + 2)/3
+        if y.shape == ():
+            return y[()]
         else:
-            if self._derivative_polynomial is None:
-                self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
-            return ((self._n+1)*self._derivative_polynomial(x) -
-                    x*self(x))/(x**2-1)
+            return numpy.squeeze(y)
```

Comment:
What is the reason for `squeeze` here?

I also think you can use `y.item()` to return the single scalar value in the if.

Reply:
Squeeze is just my usual go-to for dimension reducing. It just wasn't adequate in this particular case which also made it unnecessary.